src/app/services/file-upload.service.ts

```diff
@@ -0,0 +1,168 @@
+
+@Injectable({
+    providedIn: "root",
+})
+export class FileUploadService {
+    private authHeaders: MaybeUndefined<{
+        [header: string]: string | string[];
+    }>;
+    private uploadToken: string;
+
+    constructor(
+        private http: HttpClient,
+        private localStorageService: LocalStorageService,
+        private appConfigService: AppConfigService,
+        private navigationService: NavigationService,
+        private datasetService: DatasetService,
+        private injector: Injector,
+    ) {
+        this.authHeaders = { Authorization: `Bearer ${this.localStorageService.accessToken}` };
+    }
+
+    public uploadFile(file: File, datasetBasics: DatasetBasicsFragment): Observable<object> {
+        const uploadPrepare$: Observable<UploadPrepareResponse> = this.uploadFilePrepare(file);
+        return uploadPrepare$.pipe(
+            tap((data) => (this.uploadToken = data.uploadToken)),
+            switchMap((uploadPrepareResponse: UploadPrepareResponse) =>
+                this.prepareUploadData(uploadPrepareResponse, file),
+            ),
+            switchMap(({ uploadPrepareResponse, bodyObject, uploadHeaders }: UploadPerareData) =>
+                this.uploadFileByMethod(
+                    uploadPrepareResponse.method,
+                    uploadPrepareResponse.uploadUrl,
+                    bodyObject,
+                    uploadHeaders,
+                ),
+            ),
+            switchMap(() =>
+                this.ingestDataToDataset(
+                    {
+                        accountName: datasetBasics.owner.accountName,
+                        datasetName: datasetBasics.name,
+                    },
+                    this.uploadToken,
+                ),
+            ),
+            finalize(() => {
+                this.updatePage(datasetBasics);
+            }),
+        );
+    }
+
+    private updatePage(datasetBasics: DatasetBasicsFragment): void {
+        this.updateCache(datasetBasics);
+        this.datasetService
+            .requestDatasetMainData({
+                accountName: datasetBasics.owner.accountName,
+                datasetName: datasetBasics.name,
+            })
+            .subscribe(),
+            this.navigationService.navigateToDatasetView({
+                accountName: datasetBasics.owner.accountName,
+                datasetName: datasetBasics.name,
+                tab: DatasetViewTypeEnum.Overview,
+            });
+    }
+
+    private updateCache(datasetBasics: DatasetBasicsFragment): void {
+        const cache = this.injector.get(APOLLO_OPTIONS).cache;
+        if (cache) {
+            const datasetKeyFragment = DatasetApi.generateDatasetKeyFragment(
+                cache.identify(DatasetApi.generateAccountKeyFragment(datasetBasics.owner.id)),
+                datasetBasics.id,
+            );
+            cache.evict({
+                id: cache.identify(datasetKeyFragment),
+                fieldName: "data",
+            });
```

Comment:
I think "metadata" section should be reset too.

Reply:
Replace on `metadata`